lib/llm/providers/gemini.rb

```diff
@@ -29,21 +29,37 @@ def embed(input, **params)
       Response::Embedding.new(res.body, self)
     end
 
+    ##
+    # @see https://ai.google.dev/api/generate-content#v1beta.models.generateContent Gemini docs
+    # @param prompt (see LLM::Provider#complete)
+    # @param role (see LLM::Provider#complete)
+    # @return (see LLM::Provider#complete)
     def complete(prompt, role = :user, **params)
       params = DEFAULT_PARAMS.merge(params)
       path = ["/v1beta/models", params.delete(:model)].join("/")
       req = Net::HTTP::Post.new [path, "generateContent"].join(":")
-      messages = [*(params.delete(:messages) || []), Message.new(role.to_s, prompt)]
-      body = {
-        contents: [{
-          parts: messages.map { |m| {text: m.content} }
-        }]
-      }
+      messages = [*(params.delete(:messages) || []), LLM::Message.new(role, transform_prompt(prompt))]
+      body = {contents: [{parts: messages.map { _1.to_h[:content] }}]}
       req = preflight(req, body)
-      res = request @http, req
+      res = request(@http, req)
       Response::Completion.new(res.body, self).extend(response_parser)
     end
 
+    ##
+    # @param prompt (see LLM::Provider#transform_prompt)
+    # @return (see LLM::Provider#transform_prompt)
+    def transform_prompt(prompt)
+      if LLM::File === prompt
+        file = prompt
+        inline_data = {mime_type: file.mime_type, data: file.to_base64}
+        {inline_data:}
+      elsif String === prompt
+        {text: prompt}
+      else
+        raise TypeError, "#{self.class} does not support #{prompt.class} objects"
+      end
+    end
```

Comment:
This is good, but i think we should also support passing data as URI directly (using `file_data` and `file_uri`) and leave File for local files?
[https://ai.google.dev/gemini-api/docs/vision?lang=rest](url)

Reply:
> This is good, but i think we should also support passing data as URI directly (using `file_data` and `file_uri`) and leave File for local files? [https://ai.google.dev/gemini-api/docs/vision?lang=rest](url)

i.e if `File` then inline through base64 and if `URI` pass it directly with `file_uri`

Reply:
> This is good, but i think we should also support passing data as URI directly (using file_data and file_uri) and leave File for local files?

I think that's a separate pull request. Gemini didn't support `URI` to begin with. It should be added separately.

Reply:
Yeah. You are right! I was thinking of this PR as a vision support for Gemini :sweat_smile: